Retool how validators register and are called: everything's a tag!

thockin opened the pull request:
Looking at how to not fail silently when tags are used wrong.  I started adding more lint rules for things like "you can't use +optional on a type".  It struck me that this is something that should be handled more first-class, rather than ex post facto.  We will inevitably forget to add the lint rules for some tags.

We have the tag docs struct which includes a "contexts" list, which indicates when a tag is usable.  I tried automating that check as a lint rule and it's not ideal as-is.  It also highlighted that most of our tag docs are just WRONG (cargo culted) and that the context definitions we have are not really great.

Examples:
- optional/required should only be on types
- eachVal/eachKey claim they can be used on list values (they can't today)
- listMapKey claims it can be used on list values (I don't think that makes sense?)
- subfield claims it can be used on list values (I don't think it can)
- unionDiscriminator/unionMember claim they can be used on list values (I don't think that makes sense?)

Lastly, each ExtractValidations() call doesn't have enough information to DTRT even if we fixed the above.

This PR started with expanding the contexts to distinguish struct-fields from list/map keys/values.  It evolved into a different way of approaching tags.

Rather than each plugin's `ExtractValidations()` being responsible for checking that it is used in an appropriate context (and getting new args to indicate the extra information), I tried to do something a little more "scaffolding" centric.  It changes the init sequence to register validators which include the tag string and contexts (basically the docs struct, but more fine-grained), but also having generic code look up the tags, ensure correct context, and then call a method to produce the specific FunctionGen objects.

Basically, centralize the tag-processing.  Now generation will fail with errors like:

```
failed to generate validations: k8s.io/code-generator/cmd/validation-gen/output_tests/optional.T1: tag "k8s:optional" cannot be specified on type definitions
```

or

```
failed to generate validations: field k8s.io/code-generator/cmd/validation-gen/output_tests/options.T1.S2: tag "k8s:ifOptionDisabled": takes exactly 1 argument
```

...but the `optional` plugin didn't have to do the work.  Some plugins still need to do work, like if they only apply to strings, but that seems OK to me.

This PR includes all my commits, even when later commits undo or change part of it, so readers can see the thought process.  In particular, this centralized tag handling is insufficient for things like `union` which needs to run on the TYPE, but the tags are on FIELDs.

Now there are "tag validators" and "type validators" (which run AFTER child fields).  Union has tag handlers that are used to validate context (e.g. "+union only applies to struct fields") and accumulate info, and a type handler which uses the accumulated info.

I wanted to move registry to a different package, but to avoid a circular dep, it has to become 3 packages, and it didn't seem worthwhile.

thockin committed:
Change validation tag registry, do +optional

Instead of registering a factory-function and expecting the resulting
type to extract tags itself, we now register a "tag descriptor" which
has enough information to centrally check some contextual requirements
(e.g. where can this tag be used) and then call the plugins to process
the extracted value.  This avoid all the plugins having to do validation
for context, though they may still have to do some handling, e.g. if
they care about lists vs scalars.

This commit leaves the tree in a working state.  It only does the
`optional` tag, and subsequent commits will do more.  At the end, the
legacy code will be removed.

thockin committed:
Make +validate{True,False,Error} be regular tags

This required handling embedded tags, since those are used all over the
place.  This commit also simplifies TagContext.

thockin committed:
Get rid of old TagContext

thockin committed:
Rename TagContext2 -> TagContext

thockin committed:
Fix docs to use shared info

Also, since tag descriptors cover one tag, the new Docs() method only
needs to return one value.

thockin committed:
Clean up documentation around the tag registry

Also split tag-descriptor code to a new file

thockin committed:
Make +required and +forbidden be regular tags

thockin committed:
Make +format, +maxLength, +maxItems regular tags

thockin committed:
We don't need the TagDescriptor type assertion

thockin committed:
Make +enum be a regular tag

thockin committed:
Rename tag-validator types to all be ...Validator

It's more verbose but clearer.

thockin committed:
Cleanup: always use validations.Add

This was exposed by a bug where I ended up clobbering rather than
appending.  No reason I can see not to do it this way always.

thockin committed:
Rename TagRegistry -> ValidatorRegistry

thockin committed:
Rename TagDescriptor -> TagValidator

thockin committed:
Rename TagContext -> Context

thockin committed:
Rename TagScope -> Scope

thockin committed:
Rename TagDoc.Contexts -> TagDoc.Scopes

thockin committed:
Merge tags.go and validators.go

I thought splitting them was cleaner, but it's not.

thockin committed:
Make Init() take a config type like before

thockin committed:
Make +ifOption{En,Dis}abled be regular tags

thockin committed:
Sort tags before collecting validations

It makes more sense to sort here, rather than later (when emitting)
because:

Consider a type or field with the following comments:

```
// +k8s:validateFalse="111"
// +k8s:validateFalse="222"
// +k8s:ifOptionEnabled(Foo)=+k8s:validateFalse="333"
```

Tag extraction will retain the relative order between 111 and 222, but
333 is extracted as tag "k8s:ifOptionEnabled".  We iterate that map
(in a random order).  When it reaches the emit stage, the
"ifOptionEnabled" part is gone, and we have 3 FunctionGen objects, all
with tag "k8s:validateFalse".  They are in a non-deterministic order
because of that map iteration.  If we sort them now, we don't have
enough information to do something smart, unless we look at the args,
which are sort of opaque to us.

Sorting it earlier means we can sort "k8s:ifOptionEnabled" against
"k8s:validateFalse".  All of the records within each of those is
relatively ordered, so the result here would be to put "ifOptionEnabled"
before "validateFalse" (lexicographical is better than random).

thockin committed:
Remove older validator code, now dead

thockin committed:
Rename things to older, simpler names

thockin committed:
Remove unused parameter

thockin committed:
Hide registry impl behind an interface

thockin committed:
Clean up tag-docs, add args and usage

thockin commented:
For giggles I added some WIP commits to make `+eachVal` a "regular" tag.  To do that, I had to do listMapKey and listType, too.

This probably could have been simpler, and list-map could be treated "specially" (with info passed to validators), but I wanted to see if I could do it functional-style.  

It sorta works.  There are some issues around pointerness, still, and it's not as understandable as I'd like, but I think you can get the idea.  Basically there's a family of list-oriented validators which work together to collect info about keys and then extract old values when we can.

The big remaining problem is that tags are discovered in sort-order, which means that `+eachVal` is run before `+listMapKey`.  I see two ways to solve it:

1) internally -- collect a list of funcs to call from the new FieldValidator, which runs after all tags.

2) let tags declare a sort order for discover (they already do for execution).  E.g. "schematic" vs "content" (words TBD).

If we like the approach, we can see how subfield would work, too.

thockin commented:
For more giggles, I made `+k8s:subfield(ls)=+k8s:eachVal2=+k8s:format=ip-sloppy` work - that's running on each value of a subfield of list type.

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/registry.go:
```diff
@@ -17,10 +17,96 @@ limitations under the License.
+}
+
+func RegisterTagDescriptor(desc TagDescriptor) {
+	allTags.add(desc)
```
All of the registration is init-functions, which are single-threaded.  Actually, I looked it up and it's single-threaded WITHIN A PACKAGE.  Hmm.

thockin replied:
will fix

thockin committed:
Make +subfield a regular tag

Cleanup and tests will be in subsequent commits.

thockin committed:
Use new SnippetWriter method to propagate errors

thockin committed:
Make +eachKey a regular tag

Cleanup and tests will be in subsequent commits.

thockin committed:
Add comments to Validations

This lets us emit comments into the output file.  Use this in
optional/required when we find a non-pointer struct.

thockin committed:
Add tag tests for +optional, +required, +forbidden

a) Move tag tests to a "tags" subdir
b) Use more descriptive names
c) Add a doc_test

This covers optional/required/forbidden

thockin committed:
Add tag tests for +validate{False,True}

thockin committed:
Add tag tests for +format, +maxLength, +maxItems

This identified that calling `+k8s:format` or `+k8s:maxLength` on a
typedef'ed string did not work.

thockin committed:
Add tag tests for +enum

thockin committed:
Add tag tests for +ifOption{En,Dis}abled

thockin committed:
Distinguish TagValidators from TypeValidators

TypeValidators will be called on every type and should be comparatively
rare.  This is needed to make the union validation work.  Specifically:

1) TagValidators are triggered to accumulate details about which fields
   are in which unions, as well as about discriminators. Emit no
   validations.
2) TypeValidator emits the final validation.

It's a bit more spread out than before but it allows the centralization
of tag handling.

More renames will follow, to simplify the UX.

thockin committed:
Make +union be a regular tag

thockin committed:
Add tag tests for +union and +unionMember

jpbetz commented:
I've caught up on the commits.  I'll put some comments in the code, but would like to start with some high level thoughts:

The improvements to how validators are registered and how context scopes are checked all looks great.

Separating out TypeValidators looks fine.  Unions are the only tag we have (that I'm aware of) that needs to inspect multiple tags.

Re: Making `each*` normal tags-  The topic of using our tag discovery in other generators (like openapi, defaulting, maybe even conversion) that we talked about last week is something I'd like to pursue.  Currently, we have a nice separation between the discovery pass and the emit pass.  But the `childNode` data structure that the passes share is specialized to validation.  I like how making `each*` normal tags moves us in the direction of having less validation specific information in `childNode`.  So I'm happy with this PR in the sense that it moves us in the right general direction.

I do wonder if we're doing to end up needing a third pass to make this all clean?  Currently we have: 

`(gengo universe) -> discovery pass -> (childNode) -> validation emission pass -> (validation code)`

Where any validation specific code in `childNode` can only be moved into the emission pass as we generalize unless we add another data structure.  Should we have something more like:

`(gengo universe) -> discovery pass -> (childNode) -> validation prep pass -> (validationChildNode) -> validation emission pass -> (validation code)`

?

I don't think we should hold this PR up on this idea. But this seems like a good place to discuss the longer term goals.

thockin commented:
ACK the sentiment!  I think the intermediate that you are looking for is ultimately going to be the "iptables" version of this.  Since discussion on Friday, I have been thinking a lot about it, and I think I see how to get there, but I want to focus on delivering SOMETHING before retooling it all again.  I think this "regularity" of tags helps.

I have spent some time hacking on weekend, cleaning up the output tests for tags (found some bugs!!) and when I am done witht htat I will rip out all the now unused generator code, but in discovery and emit passes.  I think the result will make it plain how we can proceed.

jpbetz commented:
> I want to focus on delivering SOMETHING before retooling it all again. I think this "regularity" of tags helps.

+100. This is all in the right general direction.

> I will rip out all the now unused generator code, but in discovery and emit passes. I think the result will make it plain how we can proceed.

SGTM!  I'll drop in review feedback on the individual commits that look the most stable next. Future review passes should be lower latency now that I'm spun up on this PR.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:
```diff
@@ -1462,6 +1471,82 @@ func toGolangSourceDataLiteral(sw *generator.SnippetWriter, c *generator.Context
 		sw.Do("$.|private$", c.Universe.Type(types.Name(v)))
 	case *validators.PrivateVar:
 		sw.Do("$.|private$", c.Universe.Type(types.Name(*v)))
+	case validators.WrapperFunction:
```
Neat trick. The composition this offers is really great.

thockin replied:
I learned it from watching you, Joe!

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go:
```diff
@@ -0,0 +1,344 @@
+	return listTagsValidScopes
+}
+
+// LateTagValidator indicatesa that validator has to run after the listType and
+// listMapKey tags.
+func (eachValTagValidator) LateTagValidator() {}
```
Long term, I like your idea of having logical groups of tags (i.e. structurally informing types like listType being a seaprate group).  This seems fine for now until we build that out.

thockin replied:
Yeah, this is kind of dirty but it was easy and didn't impact all the other tags.

aaron-prindle commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/subfield.go:
```diff
@@ -0,0 +1,134 @@
+	subContext := Context{
+		Scope:  ScopeField,
+		Type:   submemb.Type,
+		Parent: t, //FIXME: t or context.Type?
```
+1 for `t`, I believe using the resolved underlying types here rather than aliases is preferred IIUC.

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/registry.go:
```diff
@@ -17,10 +17,104 @@ limitations under the License.
+}
+
+func (tr *TagRegistry) ExtractValidations(context TagContext2, comments []string) (Validations, error) {
+	if !tr.initialized.Load() {
```
Since it is atomic, this would mean SOMEHOW we called `ExtractValidations()` before `Init()` -- that seems like a fatal problem.  Once `initialized` is set, it is never un-set, so ultimately this is an `ASSERT()` - it should never ever fire.

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/registry.go:
```diff
@@ -17,10 +17,104 @@ limitations under the License.
+
+func (tr *TagRegistry) Docs() []TagDoc {
+	var result []TagDoc
+	for _, v := range tr.descriptors {
```
fixed in a later commit ;)

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/registry.go:
```diff
@@ -62,13 +65,17 @@ func (tr *TagRegistry) init(c *generator.Context) {
 	sort.Strings(tr.index)
 }
 
+// ExtractValidations evaluates a block comments for a given context (e.g. a type
```
fixed in a later commit

thockin commented on vendor/k8s.io/gengo/v2/generator/snippet_writer.go:
```diff
@@ -165,6 +165,21 @@ func (s *SnippetWriter) Dup(w io.Writer) *SnippetWriter {
 // Append adds the contents of the io.Reader to this SnippetWriter's buffer.
 func (s *SnippetWriter) Append(r io.Reader) error {
+	if s.err != nil {
+		return nil
```
ACK, will comment that this is the samew behavior as `Do()`

thockin commented on staging/src/k8s.io/apimachinery/pkg/api/validate/each.go:
```diff
@@ -0,0 +1,111 @@
+		if cmp != nil && len(oldSlice) > 0 {
+			p := lookup(oldSlice, val, cmp)
+			if p != nil {
+				old = *p
```
The behavior here is odd, but I didn't want to make it any more complex.

`lookup` always returns a pointer, even a `*[]string`.  The Nilable part mostly governs whether we need to take the address of the item or not.  We could probably replace with a `get func(obj []T, i int) P` argument, but that makes all the simple cases ugly.

After the benchmarking experience, I am wondering if all the effort to "always pass nilable types" is justified.  It might be much simpler code to always just pass the field-under-validation as a value -- seems like pessimization, but Go is surprising wrt allocations.

I am leaving that experiment for AFTER all this stuff lands :)

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/subfield/validatefalse/doc.go:
```diff
@@ -60,3 +63,16 @@ type T1 struct {
+
+	// +k8s:subfield2(stringField)=+k8s:validateFalse="subfield T3.T2.StringField"
+	// +k8s:subfield2(structField)=+k8s:subfield2(stringField)=+k8s:validateFalse="subfield T3.T2.StructField.StringField"
+	// +k8s:subfield2(sliceField2)=+k8s:eachVal2=+k8s:subfield2(stringField)=+k8s:validateFalse="subfield T3.T2.SliceField[*].StringField"
```
I am retooling the tests -- I have not made it to subfield yet :)

thockin committed:
Add +immutable as a regular tag

This is needed to test listmaps.

thockin committed:
Make +eachVal a regular tag

Cleanup and tests will be in subsequent commits.

thockin committed:
Fix non-tag list & map iteration, clean up listmap

Upon trying to clean up the now obsolete builti-in code, I realized I
had totally missed the non-tag iteration of lists (e.g. calling the
type-validation func for each item).  Fixing that was deeper than I
expected.

After this:
- the eachVal/eachKey validators expose a helper function for use by
  main generation code (they are special).
- discover() looks for the case of aliases to maps/slices and adds
  type-validation calls to "for each" the key/val types' validation funcs.
- discoverStruct() does the same for map/slice fields
- discoverAlias() is gone, it had been reduced to ~5 LOC and a comment.
- all built-in iteration logic is gone, in favor of the eachVal and
  eachKey tag logic
- safe.NewListMap is gone
- hasValidations() is now a global rather than only available to
  generation-phase (with hazard documented)
- the last built-in tag is gone
- Some edge cases we just don't support (because it's hard and we don't
  need them):
  - alias to pointer
  - pointer to alias of slide/map/pointer

I went through all the test differences and made sure they made sense.
In hindsight, these issues should have been obvious in earlier commits.

There's still some room for cleanup after this.

thockin committed:
Reorder args to type-validation funcs

For some historical reason, the order was different than tag validators.
This commit makes them the same so type functions can be used as regular
validators (needed to do lists and maps).

thockin committed:
Reorder arguments to internal function literals

This is more for consistency than functional reasons.

thockin committed:
Use 2 list/map elements in the test fixture

thockin committed:
Make tests for slices more robust and regular

thockin committed:
Make tests for maps more robust and regular

thockin committed:
Fix double-validation for aliases to slices/maps

Exposed by the tests in the last commits.  Consider:

This unveiled a bug with typedefs to slices.  Consider:

```
type Foo []Bar

type Struct struct {
    FooField Foo
}
```

Previously we would emit a `Validate_Foo()` function which iterated the
list and called `Validate_Bar()` for each item, and then call that for
Struct.FooField.  Then we would iterate the list because we treated it
like a regular list field.  Which it is not.

When we find an alias (typedef) we should not peek at the underlying --
the type function should handle the iteration.

Testing also exposed an issue with slices of pointers to typedefs to
slices/maps.  This one is not fixed - it's not something we do and I
didn't feel like chasing it further.

thockin committed:
Move IsDNS1123Subdomain() to pkg content

This is needed to break a circular dep in upcoming work:

package k8s.io/apimachinery/pkg/api/validate
	imports k8s.io/apimachinery/pkg/api/equality
	imports k8s.io/apimachinery/pkg/apis/meta/v1
	imports k8s.io/apimachinery/pkg/labels
	imports k8s.io/apimachinery/pkg/util/validation
	imports k8s.io/apimachinery/pkg/api/validate: import cycle not allowed

k8s.io/apimachinery/pkg/util/validation is a problem

thockin committed:
Move IsQualifiedName() to pkg content

This is needed to break a circular dep in upcoming work.

thockin committed:
Move IsLabelValue() to pkg content

This is needed to break a circular dep in upcoming work.

thockin committed:
Make tests for list-maps more robust and regular

thockin committed:
Get rid of forEachField, simpler

thockin committed:
Add tag tests for +eachVal for lists

thockin committed:
Regenerate

thockin committed:
Add tag tests for +eachVal for maps

thockin committed:
Regenerate

thockin committed:
Add tag tests for +eachVal for list-maps

thockin committed:
Regenerate

thockin committed:
Add tag tests for +eachKey for maps

thockin committed:
Regenerate

thockin committed:
Add tag tests for +subfield

thockin committed:
Regenerate

thockin committed:
Clean up old eachKey code

thockin committed:
Regenerate

thockin committed:
Clean up old eachVal code

thockin committed:
Regenerate

thockin committed:
Clean up old subfield code

thockin committed:
Clean up tag name consts

thockin committed:
If WrapperFunction has no args, just call the func

thockin commented:
OK, I think this is FINALLY working.  It spiralled a bit out of control, and I can break it into smaller PRs, but it is here for review.

Mostly you can read it commit-by-commit.  Later commits will undo some sins of previous ones, which I have left in because I am not THAT good at rebasing.  They will all squash away.

I tried to comment things for future us.

I add and refactored a ton of tests. I looked at all the final results and they seem right, but they seemed right along the way - confirmation bias is a hell of a drug.

There's one WIP commit, with a couple of FIXMEs left - those 2 need answers.  Specifically:

* What other values of `mapType` are there and what should we do with them?
* Should `listMapKey` use JSON names or Go names (or try to figure it out?)  Today we use Go names, but subfield opted for JSON.  Inconsistencies are bad, mmkay?

I have a few ideas for further cleanup, LATER.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:
```diff
@@ -235,11 +240,10 @@ type typeNode struct {
-	key         *childNode     // populated when this type is a map
-	elem        *childNode     // populated when this type is a map or slice
-	underlying  *childNode     // populated when this type is an alias
-	listMapKeys []types.Member // populated with listMapKey values when this type is listType=map slice
```
Oh wow. This makes the payoff of this refactor super obvious. :+1:

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:
```diff
@@ -412,6 +433,94 @@ func (td *typeDiscoverer) discover(t *types.Type, fldPath *field.Path) (*typeNod
+					// type has validations, otherwise it is noise.
+					if hasValidations(underlying.node) {
+						v, err := validators.ForEachVal(fldPath, underlying.childType,
+							validators.Function("iterateListValues", validators.DefaultFlags, funcName))
```
This approach is so great.

Very minor thing I'll go ahead and mention: Looks like "iterateListValues" is a placeholder tag name since we don't need one here?  Maybe comment on this? When reading the call site it took me a moment to grok what this string represented.

thockin replied:
yeah, it isn't actually used anywhere, but it was useful when debugging :)

jpbetz reviewed (commented):
I've made it through a full pass.

Nothing here really concerns me.  The tag-ification of all-the-things makes me so happy.  Test coverage improvements look solid.

A few comments around immutability, but otherwise LGTM.

Thanks a ton. This is a huge improvement.

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/openapi.go:
```diff
@@ -20,126 +20,198 @@ import (
+func (maxLengthTag) GetValidations(context TagContext, _ []string, payload string) (Validations, error) {
+	var result Validations
+
+	t := context.Type
+	if t.Kind == types.Alias {
+		t = t.Underlying
+	}
```
I think it's idiomatic gengo, and not worth encapsulating.  I know Go typedefs are only one-deep, but I still sometimes want to write this as `for t.Kind == types.Alias` :)

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/required.go:
```diff
@@ -75,51 +108,53 @@ func (requiredDeclarativeValidator) ExtractValidations(t *types.Type, comments [
+	// one does.  Since Go doesn't do partial specialization of templates, we
+	// do manual dispatch here.
+	switch context.Type.Kind {
+	case types.Slice:
```
Well, I tried to handle arrays where I could but, we don't use it and I didn't add tests for it.

Also:

* arrays can't be `nil`
* arrays have a fixed `len()`

So what would it mean for an array to be optional?  That all the values satisfy optional?  

I could add an explicit error and just say we don't support optionalness for arrays, or I could just leave it as it is, which falls through to a value comparison which (for now) just looks for the zero-value of the array, whatever the Elem type is...

Opinions welcom

jpbetz replied:
> So what would it mean for an array to be optional? 

This convinced me.  I think we're good.

thockin commented on staging/src/k8s.io/apimachinery/pkg/api/validate/union.go:
```diff
@@ -38,20 +38,24 @@ import (
+			field.InternalError(fldPath,
+				fmt.Errorf("number of field values (%d) does not match number of union members (%d)",
+					len(fieldValues), len(union.members))),
+		}
```
felt so good to get good errors

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/validators.go:
```diff
@@ -17,9 +17,149 @@ limitations under the License.
+	Docs() TagDoc
+}
+
+type TypeValidator interface {
```
I did add a godoc in a later commit

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/multiple_tags/doc.go:
```diff
@@ -25,25 +25,33 @@ import "k8s.io/code-generator/cmd/validation-gen/testscheme"
 
 var localSchemeBuilder = testscheme.New()
 
+// validateTrue should sort after validateFalse, but all the validateFalse
```
Good call!  Insert that text pretty much verbatim.

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/main.go:
```diff
@@ -118,6 +118,29 @@ func printDocs() {
 	for i := range docs {
 		d := &docs[i]
 		slices.Sort(d.Scopes)
+		if d.Usage == "" {
```
I waffled over whether to use omitEmpty or not -- opinions welcome

thockin commented on staging/src/k8s.io/apimachinery/pkg/api/validate/immutable.go:
```diff
@@ -0,0 +1,52 @@
+// the caller needs to compare types that are not comparable, they should use
+// ImmutableNonComparable instead.
+func Immutable[T comparable](opCtx operation.Context, fldPath *field.Path, value, oldValue *T) field.ErrorList {
+	if value == nil || oldValue == nil {
+		return nil
+	}
```
Good question!  Optional has so many fun cases.

We could make some tags "Early" the same way some are "Late", and run them before short-circuits.  That feels kindy or whack-a-mole.  

We could make `+optionalImmutable`.

We could make optional be aware of immutable and check for its existence.

We could .... ideas?

thockin replied:
Another case I was handling wrong with immutable: if the elements of a list-map are immutable, you can't remove an item from the list (my testcase was wrong).  Can you add items to a list in that case?

jpbetz replied:
> We could make +optionalImmutable.

Yeah, we might need a few flavors of immutable for optional fields.

Flavor 1: The field can't transition from unset->set or set>unset.
Flavor 2: The field can't be unset once set.  But so long as the field is unset, it can be set whenever.
Flavor 3: The field can be set or unset whenever, but it cannot be modified while set.

BTH, I feel like Flavor 3 is utterly useless. It's easy to bypass and so offers no meaningful guarantees.

Flavor 1 is the strongest form of immutability.  So if we allow `+immutable` on optional fields, this is probably what I wan't it to mean.

Flavor 2 SEEMS really handy. It's more of a state transition rule type thing... Say you have an optional deletion timestamp field.  It could be useful to have it be immutable-once-set.  I might even want to call it `+immutable-once-set`?

> I think it's like a one-way door. You can add items but not remove them - agree or disagree?

Append only lists/maps/listMaps seem useful, and feel like a flavor of immutability. I think we might need them. 

I believe for [PodSpec.Containers](https://github.com/kubernetes/kubernetes/blob/f422a58398d457a88ac7b05afca85a0e6b805605/staging/src/k8s.io/api/core/v1/types.go#L3813) we need full immutability.  So I'd expect `+immutable` to be sufficient for that containers field.

thockin replied:
> Flavor 1 is the strongest form of immutability.

I have a pending patch to this PR to implement this semantic for +immutable, with the open question of slices/maps

> Flavor 2 SEEMS really handy. It's more of a state transition rule type thing
> I'd be happy to wait for an actual use case to emerge before implementing this)

100% agree - no room for speculation at this stage.  Also agree that a different tag seems right.

> Append only lists/maps/listMaps seem useful, and feel like a flavor of immutability. I think we might need them.

When you say it this way, it emphasizes that there's both parent (the list) and child (the item in the list) implications.

> I believe for PodSpec.Containers we need full immutability

To be clear, not immutability of the whole list (which is recursive and seems obvious) but immutability of the set of list-map keys - no adding or removing items.  That's not really what `+k8s:eachVal=+k8s:immutable` means either.

So I still need a conclusion on:

Given a slice, what does `+k8s:eachVal=+k8s:immutable` mean?
* for non-listmap lists, it means nothing because updates don't correlate them (perhaps they should correlate by index?)
* for listmaps, it clearly means that correlated items may not be modified, but does it also mean:
a) items may be added
b) items may be removed
c) a && b
d) neither (effectively meaning the same thing as `+k8s:immutable` on the slice field

We could just look at the context.ParentType and throw an error if you use it on a slice, and defer the problem.  This throws a wrench into my update tests - immutable was how I was testing listmap correlation -- any ideas for a simple update validation to prove this?

thockin replied:
I guess I could add `validateFalseUpdate` where the error string indicted (old | new) x (specified | nil) ?

Speaking of testing, should we kill off validateTrue?

jpbetz replied:
> > I believe for PodSpec.Containers we need full immutability
> 
> To be clear, not immutability of the whole list (which is recursive and seems obvious) but immutability of the set of list-map keys - no adding or removing items. That's not really what `+k8s:eachVal=+k8s:immutable` means either.

Oooh.  I hadn't groked this when responding before.

> So I still need a conclusion on:
> 
> Given a slice, what does `+k8s:eachVal=+k8s:immutable` mean?
> 
> * for non-listmap lists, it means nothing because updates don't correlate them (perhaps they should correlate by index?)
> * for listmaps, it clearly means that correlated items may not be modified, but does it also mean:
>   a) items may be added
>   b) items may be removed
>   c) a && b
>   d) neither (effectively meaning the same thing as `+k8s:immutable` on the slice field

If I had to guess what users would expect, I would guess (d).

> We could just look at the context.ParentType and throw an error if you use it on a slice, and defer the problem. This throws a wrench into my update tests - immutable was how I was testing listmap correlation -- any ideas for a simple update validation to prove this?

Maybe we need something like `+k8s:immutableKeys`  or `+k8s:keyUpdateRule=(exact|appendOnly)` (I can't think of a non-gross way to express it yet..).

thockin replied:
Had a little time to think about this more, no conclusion but some thoughts.

Whether a list field can have items added/removed should be a tag of its own.  For example, we can assume that lists, by default, are mutable (wrt `len()`) or not, that's a decision of which is more common.

`+k8s:immutable` for lists is pretty obviously inclusive of not allowing add/remove.

Something like `+k8s:eachVal=+k8s:format=dns-label` pretty obviously applies to all elements of the "new" list, and totally ignores the old list entirely.

Something like `+k8s:eachVal=+k8s:immutable` seems like it should apply to each item of the new list, with the corresponding old value (or nil), but that seems allergic to add/remove.  

This feels contrived - I doubt if anyone wants immutable values in a list but add/remove are OK.

Aside: given a list-map field in which an item has been removed, should we call validators with `value=nil` and `oldValue=<value>` ?  Or do we assert that `value` is always expected to be valid?

jpbetz replied:
> Whether a list field can have items added/removed should be a tag of its own. 

This is where I landed as well.

> This feels contrived - I doubt if anyone wants immutable values in a list but add/remove are OK.

+1

> Aside: given a list-map field in which an item has been removed, should we call validators with `value=nil` and `oldValue=<value>` ? Or do we assert that `value` is always expected to be valid?

I think set->unset transitions are needed for `+k8s:immutable` so we can declare that once the field is set, it's locked-in and you can't unset it and set it to a different value.

We had a primordial form of update validation where the tag had to opt-in to observe set->unset transitions.  I think that would work if we don't want to force validation authors with dealing with this edge case. We ended up going down the path of calling all validations uniformly, which also has some nice properties, but requires dealing with these subtlties.

thockin replied:
I am AFK, but thinking about it. Don't have code in front of me play with. 

If you buy that a list's add / remove behavior is governed by a different tag, not really discussed here, then it seems okay that +eachVal would apply to all elements of `value`, including correlated `oldValue` data, if the key existed.

Specifying `+eachVal=+immutable` on a list, without restricting add / remove, would call immutable with potentially nil oldVal (in which case it would fail the immutability check).  That seems correct, doesn't it?  If I was calling a simple string format validator, that's what I would want.  

For my test cases, this does what I need which is to prove that the correlation works for fields that are common in old and new, and that it produces nil for elements that are not in old.

I feel like this is too obvious. Was I just making it hard before, or am I missing something now? Question?

jpbetz replied:
> Specifying `+eachVal=+immutable` on a list, without restricting add / remove, would call immutable with potentially nil oldVal (in which case it would fail the immutability check). That seems correct, doesn't it? If I was calling a simple string format validator, that's what I would want.

Yes, this is exactly what I'd expect.

> For my test cases, this does what I need which is to prove that the correlation works for fields that are common in old and new, and that it produces nil for elements that are not in old.
> 
> I feel like this is too obvious. Was I just making it hard before, or am I missing something now?

It's twisty!  But our primitives are composing, so I'm putting this one in the win column :)

jpbetz replied:
^ @yongruilin or @aaron-prindle We should document the above very carefully.  Anyone interested in opening an issue to track and adding some notes?  In particular, the semantics of `+immutable` on different types as well as on `+eachVal=+immutable` should probably be tabulated out.  I think I'm convinced the semantics can be derived from principals but for the beneift of our users, spelling it out in documentation seems appropriate.

yongruilin replied:
Based on the conversation, I created https://github.com/jpbetz/kubernetes/issues/95. PTAL and see whether that is align with the consensus.

thockin commented on staging/src/k8s.io/apimachinery/pkg/api/validate/immutable.go:
```diff
@@ -0,0 +1,52 @@
+// provided. Unlike Immutable, this function can be used with types that are
+// not comparable, at the cost of performance.
+func ImmutableNonComparable[T any](opCtx operation.Context, fldPath *field.Path, value, oldValue T) field.ErrorList {
+	if !equality.Semantic.DeepEqual(value, oldValue) {
```
Oh.  Good catch!  I think they should both be nil or neither, agree?

thockin replied:
Actually...

Value should never be nil, but oldval can be.  We don't really have "update-only" validations (maybe we need to codify that) so on a CREATE we will find value is valid but oldValue is nil.  But on an update of a pointer field we would see the same thing.  That's not good.

We need to check the opCtx.  Then we can assert the same nil-ness.

thockin replied:
fixed

thockin commented:
@jpbetz I also will direct your attention to FIXMEs:

[one](https://github.com/jpbetz/kubernetes/pull/89/commits/ae444e01f75e9831b8226b1566268ef3f12a9553#diff-928d1a3f74fa39408ef20be4d064ded6a0ada5cf5c6226229223418e06ff219eR72-R83)

[two](https://github.com/jpbetz/kubernetes/pull/89/commits/ae444e01f75e9831b8226b1566268ef3f12a9553#diff-928d1a3f74fa39408ef20be4d064ded6a0ada5cf5c6226229223418e06ff219eR143-R151)

thockin committed:
Fix lint to use proper +default tag

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go:
```diff
@@ -0,0 +1,345 @@
+		return Validations{}, fmt.Errorf("can only be used on list types")
+	}
+
+	if payload == "map" {
+		if realType(t.Elem).Kind != types.Struct {
+			return Validations{}, fmt.Errorf("only lists of structs can be list-maps")
+		}
+
+		// Save the fact that this list is a map.
+		if lttv.byFieldPath[context.Path.String()] == nil {
+			lttv.byFieldPath[context.Path.String()] = &listMap{}
+		}
+		lm := lttv.byFieldPath[context.Path.String()]
+		lm.declaredAsMap = true
+	} // FIXME:else?
```
Supported values: atomic/set/map

For "atomic" and "set" not additional validation or tracking of information is required.

If the payload is a unrecognized string, we should report and error I think.

thockin replied:
So say we all.  In next push.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go:
```diff
@@ -0,0 +1,345 @@
+	}
+
+	found := false
+	for _, memb := range realType(t.Elem).Members {
+		//FIXME: should this be the JSON name?  That's a change...
+		if memb.Name == payload {
+			if k := realType(memb.Type).Kind; k != types.Builtin {
+				return Validations{}, fmt.Errorf("only primitive types can be list-map keys, not %s", k)
+			}
+			found = true
+			break
+		}
```
Yes, we should use JSON name here. This will become important later when we deal with OpenAPI publication.

thockin replied:
all the existing keys are defined as Go names -- should I just try to use either one?  If I can't find a JSON name, use the Go?  Or maybe just make sure it hard-fails, so authors need to do it right?

thockin replied:
Changed it to just error if the JSON name is not found.

thockin committed:
Test scheme: update args in the wrong order

thockin commented:
It's GREEN!!!

I simply accepted that immutable on a list, in the face of updates which add an item, will fail.  It can't be immutable if the old value was not specified.  To be clear, I think this is BEYOND a corner case, but it's useful for testing.

thockin commented:
Also 3 gengo PRs pending

thockin committed:
Update gengo dep to 20250207200755-1244d31929d7

jpbetz commented:
LGTM

I'm all caught up.  This is excellent progress.  Let's iterate on the dev branch from here

caozhuozi commented:
Hi, sry to borther! I came across this PR and I'm interested in this topic. Is there any KEP that provides background information on it?

yongruilin commented:
> Hi, sry to borther! I came across this PR and I'm interested in this topic. Is there any KEP that provides background information on it?

Hi @caozhuozi , you can check on https://github.com/kubernetes/enhancements/pull/5074 . Thanks for your interests.

caozhuozi commented:
@yongruilin Thank you!!
